### ISSUE TITLE
$oidc disappears in argocd-cm ConfigMap

### ISSUE THREAD

**Issue by eli-halych** [argoproj/argo-cd#8635]
When I deploy Argo CD, I want this key-value pair in argocd-cm ConfigMap: `oidc.config.clientSecret: $oidc.keycloak.clientSecret` (according to [THIS](https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/keycloak/) guide).

However, after I deploy Argo CD through the GitLab Pipeline, I get this instead: `oidc.config.clientSecret: .keycloak.clientSecret`.

`$oidc` disappears. How can I make sure that `$oidc.keycloak.clientSecret` is deployed as it is? This makes it impossible for me to use clienSecret.

**Comment by crenshaw-dev** [comment]
I'm not super familiar with GitLab pipelines. How are you deploying Argo CD in pipelines? Are you using kustomize/kubectl in some container step?

Kinda feels like something is interpreting `$oidc` as an empty variable. But without knowing all the tools in the chain, I'm not sure which tool might be doing that.

**Comment by eli-halych** [comment]
@crenshaw-dev 
I am deploying it using Helmsman and a Helmchart. The pipeline runs `helm` and `helmsman` commands. 

```bash
user@server:~$ helm version
version.BuildInfo{Version:"v3.7.0", GitCommit:"eeac83883cb4014fe60267ec6373570374ce770b", GitTreeState:"clean", GoVersion:"go1.16.8"}
```

**Comment by crenshaw-dev** [comment]
Can you run `helm template` on the sources manually, just to make sure it's preserving the `$oidc`?

What tool are you using to confirm that `$oidc` is gone? Maybe there's a bug in some presentation software. I'd personally trust output from `kubectl`.

**Comment by eli-halych** [comment]
1. `helm template .` output:
```yaml
clientSecret: $oidc.keycloak.clientSecret
```
`$oidc` is not gone.

2. I used `kubectl edit argocd-secret -n argocd` and this is where I saw that the `$oidc` was gone.

**Comment by eli-halych** [comment]
I couldn't manually do it with Helmsman. Also, it looks like that the dollar sign, or the whole `$oidc` part, needs to be escaped to not be treated as a variable with null value. I tried \$, $$, $(oidc.keycloak.clientSecret), %$, but nothing worked.

**Comment by bigwheel** [comment]
$$oidc ?
I thought this issue might be related to https://github.com/argoproj/argo-cd/issues/6878 .

**Comment by kompiz** [comment]
> $$oidc ? I thought this issue might be related to #6878 .

Yes! Thank you! `$$oidc` in the argocd-app.yaml renders `$oidc!`

**Comment by byn-sanket-godbole** [comment]
> > $$oidc ? I thought this issue might be related to [#6878](https://github.com/argoproj/argo-cd/issues/6878) .
> 
> Yes! Thank you! `$$oidc` in the argocd-app.yaml renders `$oidc!`

I am facing the same issue here,
but even when I use $$ like this `clientSecret: $$dex-secret:clientSecret` this $$ is getting replaced with `clientSecret: 24022dex-secret:clientSecret`
I dont understand from where these numbers are coming from
